File: Sources/ExcalidrawRender/ElementDrawable.swift
import ExcalidrawMath
import ExcalidrawModel
import Foundation
import RoughKit

/// Generates the rough.js `Drawable` for an element, in element-local
/// coordinates (origin at the element's `x, y`). The scene renderer applies the
/// element's translation and rotation when drawing. Ports the dispatch in
/// `_generateElementShape` (`packages/element/src/shape.ts`).
///
/// Scope: rectangle, diamond, line, arrow this increment. Ellipse, freedraw and
/// text shapes are the next Phase 2 increment.
public enum ElementDrawable {
    private static let generator = RoughGenerator()

    public static func drawable(for element: ExcalidrawElement) -> Drawable? {
        let o = RoughOptionsBuilder.options(for: element)
        let w = element.base.width
        let h = element.base.height

        let rounded = element.base.roundness != nil

        switch element.kind {
        case .rectangle, .embeddable, .iframe:
            if rounded {
                return generator.polygon(roundedRectanglePoints(w, h), options: o)
            }
            return generator.rectangle(x: 0, y: 0, width: w, height: h, options: o)
        case .diamond:
            // Diamond vertices at the midpoints of each bounding-box edge.
            let pts = [Point(w / 2, 0), Point(w, h / 2), Point(w / 2, h), Point(0, h / 2)]
            return generator.polygon(pts, options: o)
        case .ellipse:
            return generator.ellipse(x: w / 2, y: h / 2, width: w, height: h, options: o).drawable
        case let .line(props):
            return linearDrawable(props.points, closed: props.polygon || isLoop(props.points), rounded: rounded, o)
        case let .arrow(props):
            return linearDrawable(props.points, closed: false, rounded: rounded, o)
        default:
            return nil // freedraw / text / image / frame: handled directly by the renderer / later increments
        }
    }

    private static func linearDrawable(_ points: [Point], closed: Bool, rounded: Bool, _ o: RoughOptions) -> Drawable? {
        guard points.count >= 2 else { return nil }
        if rounded, points.count > 2 {
            return generator.curve(points, options: o)
        }
        return closed ? generator.polygon(points, options: o) : generator.linearPath(points, options: o)
    }

    private static func isLoop(_ points: [Point]) -> Bool {
        guard points.count >= 3, let first = points.first, let last = points.last else { return false }
        return first.distance(to: last) <= 40
    }

    /// Closed outline of a rounded rectangle (`w × h`), with each corner sampled
    /// as a quarter-arc, for the hand-drawn rounded look.
    static func roundedRectanglePoints(_ w: Double, _ h: Double, samples: Int = 6) -> [Point] {
        let r = min(min(w, h) / 4, 32)
        guard r > 0 else { return [Point(0, 0), Point(w, 0), Point(w, h), Point(0, h)] }
        var points: [Point] = []
        // Corner centres and arc start angles, clockwise from the top-left.
        let corners: [(cx: Double, cy: Double, start: Double)] = [
            (r, r, Double.pi), // top-left
            (w - r, r, 1.5 * Double.pi), // top-right
            (w - r, h - r, 0), // bottom-right
            (r, h - r, 0.5 * Double.pi) // bottom-left
        ]
        for corner in corners {
            for k in 0 ... samples {
                let a = corner.start + (Double.pi / 2) * Double(k) / Double(samples)
                points.append(Point(corner.cx + r * cos(a), corner.cy + r * sin(a)))
            }
        }
        return points
    }
}

File: web/packages/render/src/element-drawable.ts
import type { ExcalidrawElement, LocalPoint } from "@cyberdynecorpai/model";
import rough from "roughjs";
import type { Drawable, Options } from "roughjs/bin/core";

const generator = rough.generator();

function isLoop(points: LocalPoint[]): boolean {
  if (points.length < 3) return false;
  const first = points[0]!;
  const last = points[points.length - 1]!;
  return Math.hypot(first[0] - last[0], first[1] - last[1]) <= 40;
}

/**
 * Closed outline of a rounded rectangle (`w × h`), each corner sampled as a
 * quarter-arc, for the hand-drawn rounded look. (parity: ElementDrawable.swift)
 */
export function roundedRectanglePoints(w: number, h: number, samples = 6): LocalPoint[] {
  const r = Math.min(Math.min(w, h) / 4, 32);
  if (r <= 0) {
    return [
      [0, 0],
      [w, 0],
      [w, h],
      [0, h],
    ];
  }
  const corners: [number, number, number][] = [
    [r, r, Math.PI],
    [w - r, r, 1.5 * Math.PI],
    [w - r, h - r, 0],
    [r, h - r, 0.5 * Math.PI],
  ];
  const points: LocalPoint[] = [];
  for (const [cx, cy, start] of corners) {
    for (let k = 0; k <= samples; k++) {
      const a = start + ((Math.PI / 2) * k) / samples;
      points.push([cx + r * Math.cos(a), cy + r * Math.sin(a)]);
    }
  }
  return points;
}

function linearDrawable(
  points: LocalPoint[],
  closed: boolean,
  rounded: boolean,
  o: Options,
): Drawable | null {
  if (points.length < 2) return null;
  if (rounded && points.length > 2) return generator.curve(points, o);
  return closed ? generator.polygon(points, o) : generator.linearPath(points, o);
}

/**
 * Generate the rough.js `Drawable` for an element, in element-local coordinates
 * (origin at the element's `x, y`). (parity: ElementDrawable.swift)
 */
export function elementDrawable(el: ExcalidrawElement, options: Options): Drawable | null {
  const w = el.width;
  const h = el.height;
  const rounded = el.roundness !== null;

  switch (el.type) {
    case "rectangle":
    case "embeddable":
    case "iframe":
      return rounded
        ? generator.polygon(roundedRectanglePoints(w, h), options)
        : generator.rectangle(0, 0, w, h, options);
    case "diamond":
      return generator.polygon(
        [
          [w / 2, 0],
          [w, h / 2],
          [w / 2, h],
          [0, h / 2],
        ],
        options,
      );
    case "ellipse":
      return generator.ellipse(w / 2, h / 2, w, h, options);
    case "line":
      return linearDrawable(el.points, el.polygon || isLoop(el.points), rounded, options);
    case "arrow":
      return linearDrawable(el.points, false, rounded, options);
    default:
      return null;
  }
}
